### PR TITLE
SCRD-4282 Implement a basic version of cross-repo depends-on

### PR DESCRIPTION
**Pull request by jhesketh**
This set of commits implements a version of cross repo dependencies for building changes from gerrit in OBS.

Both formats of Depends-On work meaning you can use the gerrit change id (Iaaa..ff) or a URL to the change itself.

This only works for dependencies within our gerrit. It does not allow cross-gerrit or github dependencies.

### CODE REVIEW

**Comment by jhesketh** [comment]
@stefannica It looks like I missed a part (commit amended). I've set that job to rebuild: https://ci.suse.de/job/cloud-ardana-testbuild-gerrit-test/15/

**Comment by jhesketh** [comment]
> I think this will work, it would really help if the contained a high level overview of what is going on.

Any suggestions where I should put that? Or do you just mean better code comments?

> What happens with circular depends-on. Simple case a review depends on it self.

It should be handled. As it processes each change it puts it into the changes dict with the gerrit id as the key. It also adds any dependencies to be processed to the change_ids, and if a dependency is encountered that is already in the changes dict it is skipped.


> What happens with branching, what if the depends on change id does not have the branch the source review is against.

This is a good point. It should only be an issue when using a depends-on within the same repository. Ideally this shouldn't be done as you should push your commits in order to gerrit. However, if it does happen, the branch from the first change that is processed is used. The other changes are then `git fetch`ed and `git merge`d onto that. So in theory it'll work, but it'd also be a mess and possibly incorrect.

Because we aren't doing strict gating, this may be acceptable. Otherwise we could try and catch this and fail the build.

> Are cross branch depends-on supported? Are there limits to the context in which they are supported.

As above, it should work across different repos. Depends-on within the same repo will likely give odd results.

> What happens if a change ID has more than one review associated with it it?

Good question. At the moment I think it'll fail. We should handle this appropriately.

This most commonly occurs with cherry-picks across branches. This means we'll need to track the branch state more closely and likely make some assumptions that if multiple changes with the same ID exist you likely want the one on the same branch.

> What happens if all reviews associated with a change ID are already in the branch. (review state merged). If did not see where the review state was checked and I don't think I saw where the branch was checked to see if the change was already present.

So we could add some extra checks around this, but as far as I know it should be okay as is. Specifically it will perform a `git merge` on a sha that is already in the tree. This should be a safe operation and git will do nothing. (Please correct me if I'm wrong)

> I don't think any of these should gate this change which I think will do what it intends, but it would be good to have any limitations noted in the file. Since that will help in debugging failures. Which can get very tricking with depends-on. I have seen depend-on chains exceeding twenty change ids. Which make finding issues very tricky.

Okay, I think this comes back to better comments.

Could you please refer me to a large depends-on change to help with my testing?

Thanks,
Josh

**Comment by rtamalin** [comment]
My suggestion would be to add some explanatory documentation to the code, e.g. a docstring at the top of the file explaining what the intend of the script is, and and a high level overview of how the script is intended to operate.

**Comment by rtamalin** [comment]
Also we probably should start to make a clear distinction between a Gerrit "change", hereafter referred to as a Gerrit patch#, which is a decimal number identifying a specific review patch set in Gerrit, vs a Gerrit "Change ID", which is a hexadecimal hash value preceeded by a capital "I".

A  gerrit "Change ID" represents a set of "equivalent" patches across various branches; equivalence here being that they achieve the same desired effect for the code stream they are found in. And since a Depends-On reference in a commit message, if we are emulating the Zuul mode of operation, is usually a Gerrit "Change ID", and not a patch#, when we lookup a Gerrit "Change ID" we should really be looking for the patch set associated with the branch we are working on.

This distinction is especially important for a number of reasons:

- A patch set for the wrong branch may not apply cleanly and we would get a merge failure
- A patch set for the wrong branch may be missing additional changes needed in the current branch to operate correctly
- Or similarly the wrong patch set may contain additional changes that that cause a failure when used with the current branch
- A patchset may not exist for the current branch, which may indicate an error/omission on someones part, e.g. cherry-picked a dependent patch without also cherry-picking patches it depends on, or removing unnecessary Depends-On references. For the Gozer/Zuul CI I'm not sure how this is handled... @brett060102 may have a better idea of how this causes grief, but it usually causes problems for gating if not check runs... Personally I think it might be better to have a clear failure if we detect a Depends-On reference that cannot be satisfied for the current branch, as that would force us to recognise that we need to either (a) cherry-pick in another change to this branch or (b) don't in fact need the references patches for this branch
- Patch sets for different branches may have different Depends-On chain references, e.g. I cherry-pick a patch from master to stable/pike to pull in a fix for an issue that has been found... But the patch on master has an implicit dependency on some patch that has already been merged to master, but didn't need to be cherry-picked back to stable/pike for whatever reason, so now I need to cherry-pick both patches back, and add the appropriate Depends-On entry to my first patch to validate it in conjunction with the dependent patch...

I'm sure we can come up with more scenarios but that should cover most of the problem scenarios I have seen working with Gerrit/Zuul for HOS/Ardana and with upstream OpenStack over the past few years...

Also as @brett060102 pointed out we shouldn't be trying to apply changes that have already merged to the current branch, as that could either result in merge failures or really strange/subtle issues if subsequent patches on the current branch allowed it to merge again.

So in summary we need to keep in mind the following concepts:

- a Gerrit "Change ID" identifies a set of one or more Gerrit "Patch Sets"
- a Gerrit "Patch Set" identifies the set of specific changes, and possible Depends-On references for a specific branch
- a Depends-On reference should only be resolved to a patch set in the context of the current branch, and should be skipped if we determine the referenced patch set has already been merged to the current branch.

**Comment by rtamalin** [comment]
Note that the proposed change is a move in the right direction, so I'm not saying we need to fix everything in it... Just that we have a ways to go yet. ;-)

But this patch is definitely a step in the right direction... Maybe just add some more detailed comments as to how it is intended to operate for now...

File: scripts/jenkins/ardana/gerrit/build_test_package.py [hunk 1]
```diff
@@ -161,33 +171,49 @@ def cleanup(testproject):
...
 
 
 if __name__ == '__main__':
-    main()
+    change_ids = os.environ['gerrit_change_ids'].split(',')
+    develproject = os.environ['develproject']
+    homeproject = os.environ['homeproject']
+    repository = os.environ['repository']
+
+    main(change_ids, develproject, homeproject, repository)
```
Comment (JanZerebecki) [inline scripts/jenkins/ardana/gerrit/build_test_package.py]: Please don't put anything besides one function call into ``__main__``. E.g. by renaming main as it is now to build_test_packages and moving the above to a new main.

**Comment by JanZerebecki** [comment]
Nothing in this comment needs to happen in this PR.

Note that Zuul prevents the merge of a commit where any Depends-On targets are not yet merged. This implementation lacks that and thus allows merging of changes without their dependencies.

> And since a Depends-On reference in a commit message, if we are emulating the Zuul mode of operation, is usually a Gerrit "Change ID", and not a patch#

While that was true in Zuul 2.5, newer versions prefer URLs to individual pull requests / patch sets / changes (which are all specific to one branch) and Change-Ids are deprecated (see https://zuul-ci.org/docs/zuul/user/gating.html#cross-project-dependencies ). Adopting this would allow supporting this feature also for Github PRs and OBS submit requests.

The implementation before and after this PR fails for Change-Ids that are on multiple branches:
```
$ curl 'https://review.openstack.org/changes/I32328c5206b9cd6fb8d8764c079f22b6ea8bfa9e/?o=CURRENT_REVISION&o=CURRENT_COMMIT'
Multiple changes found for I32328c5206b9cd6fb8d8764c079f22b6ea8bfa9e
$ curl 'https://review.openstack.org/changes/I32328c5206b9cd6fb8d8764c079f22b6ea8bfa9e/?o=CURRENT_REVISION'
Multiple changes found for I32328c5206b9cd6fb8d8764c079f22b6ea8bfa9e
```

If we need to support the old syntax then we can use a slightly different API (returns a list instead of one item):
https://review.openstack.org/changes/?q=I32328c5206b9cd6fb8d8764c079f22b6ea8bfa9e&o=CURRENT_REVISION&o=CURRENT_COMMIT

**Comment by jhesketh** [comment]
I'm still working on this. I've just pushed up where I'm up to with the refactor which is currently untested but I'm finishing for the day. I'm hoping I'll have time to test it later tonight to get it into a good intermediate state before I address all of the feedback (thanks for which btw :-))

**Comment by jhesketh** [comment]
This is ready for re-review now. Please note the updated PR message.

**Comment by jhesketh** [comment]
Yep, you can see one here: https://ci.suse.de/job/cloud-ardana-testbuild-gerrit/56/console
